### PR TITLE
DAOS-12705 test: Ensure daos_server processes are killed between tests

### PR DESCRIPTION
**Pull request by phender**
In between running functional test files, ensure that any left over daos_server, daos_engine, or daos_agent processes are killed.

Provide more detailed logging of killing processes during testing.

Fix running extra format commands during server startup and enable functions for other tests.

Skip-unit-tests: true
Test-tag: pr test_agent_failure_isolation ServerRankFailure

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/cart/util/cart_utils.py [hunk 1]
```diff
@@ -153,22 +153,16 @@ def check_files(self, glob_pattern, count=1, retries=10):
     def cleanup_processes(self):
         """Clean up cart processes, in case avocado/apricot does not."""
         error_list = []
-        localhost = NodeSet(socket.gethostname().split(".")[0])
+        localhost = get_local_host()
         processes = r"'\<(crt_launch|orterun)\>'"
+        negative_filter=r"'\<(grep|defunct)\>'"
```
Comment (daltonbohning) [inline src/tests/ftest/cart/util/cart_utils.py:158]:
```suggestion
        negative_filter = r"'\<(grep|defunct)\>'"
```

File: src/tests/ftest/launch.py [hunk 2]
```diff
@@ -62,7 +62,8 @@
         ("tcp", "ofi+tcp"),
     ]
 )
-PROCS_TO_CLEANUP = ["cart_ctl", "orterun", "mpirun", "dfuse"]
+PROCS_TO_CLEANUP = [
+    "daos_server", "daos_engine", "daos_agent", "cart_ctl", "orterun", "mpirun", "dfuse"]
```
Comment (daltonbohning) [inline src/tests/ftest/launch.py:66]: And so continues the list... :)

File: src/tests/ftest/util/command_utils.py [hunk 3]
```diff
@@ -570,6 +570,28 @@ def run(self, raise_exception=None):
...
+        """
+        # Set up the full command by setting each specified sub-command
+        full_command = self
+        if sub_command_list is not None:
+            for sub_command in sub_command_list:
+                full_command.set_sub_command(sub_command)
+                full_command = full_command.sub_command_class
+
+        # Update any argument values for the full command
+        self.update_params(**kwargs)
```
Comment (daltonbohning) [inline src/tests/ftest/util/command_utils.py:593]: I like this. One implicit improvement to using `update_params` is that it always displays the updated value

File: src/tests/ftest/util/general_utils.py [hunk 4]
```diff
@@ -841,56 +841,53 @@ def dump_engines_stacks(hosts, verbose=True, timeout=60, added_filter=None):
...
-            "then sudo /usr/bin/pkill --signal KILL {}".format(pattern),
-            "fi",
-            "fi",
-            "fi",
-            "exit $rc",
-        ]
-        result = pcmd(hosts, "; ".join(commands), verbose, timeout, None)
-    return result
+        log.debug("No processes found on %s that match: %s", result.failed_hosts, pattern_match)
+    return kill_attempted
```
Comment (daltonbohning) [inline src/tests/ftest/util/general_utils.py:890]: I'd recommend moving this else above. Something like:
```
if not result.passed_hosts:
    log.debug("No processes found on %s that match: %s", result.failed_hosts, pattern_match)
    return False

# Now  don't need another indentation, or the `kill_attempted` variable

File: src/tests/ftest/util/job_manager_utils.py [hunk 5]
```diff
@@ -251,10 +251,9 @@ def stop(self):
     def kill(self):
         """Forcibly terminate any job processes running on hosts."""
         regex = self.job.command_regex
-        result = stop_processes(self._hosts, regex)
-        if 0 in result and len(result) == 1:
-            self.log.info(
-                "No remote %s processes killed (none found), done.", regex)
+        result = stop_processes(self.log, self._hosts, regex)
+        if result.all_failed:
```
Comment (daltonbohning) [inline src/tests/ftest/util/job_manager_utils.py:255]: `stop_processes` returns a `bool`, not result object

File: src/tests/ftest/util/server_utils.py [hunk 6]
```diff
@@ -317,51 +313,66 @@ def prepare_storage(self, user, using_dcpm=None, using_nvme=None):
...
-            if not result.passed:
+            if not self.nvme_prepare(target_user=user, ignore_config=True).passed:
                 raise ServerFailed("Error preparing nvme storage")
 
+    def scm_prepare(self, **kwargs):
+        """Run daos_server scm prepare on the server hosts.
+
+        Raises:
+            RemoteCommandResult: a grouping of the command results from the same hosts with the same
+                return status
```
Comment (daltonbohning) [inline src/tests/ftest/util/server_utils.py:329]: Based on feedback from other test team members, I'd recommend adding a docstring for **kwargs. Similar to
https://github.com/daos-stack/daos/blob/5f50ba88f7bf12a5c1a67a3f1349cc3533fafca5/src/tests/ftest/util/test_utils_container.py#L916-L918

File: src/tests/ftest/cart/util/cart_utils.py (hunk 1 above)
Comment (phender) [reply src/tests/ftest/cart/util/cart_utils.py:158]: Missed that one, thanks.

File: src/tests/ftest/util/command_utils.py (hunk 3 above)
Comment (phender) [reply src/tests/ftest/util/command_utils.py:593]: In this case it's the handling of an unexpected argument

File: src/tests/ftest/util/job_manager_utils.py (hunk 5 above)
Comment (phender) [reply src/tests/ftest/util/job_manager_utils.py:255]: I forgot to save the change, doh!

File: src/tests/ftest/util/server_utils.py [hunk 7]
```diff
@@ -324,6 +323,10 @@ def prepare_storage(self, user, using_dcpm=None, using_nvme=None):
     def scm_prepare(self, **kwargs):
         """Run daos_server scm prepare on the server hosts.
 
+        Args:
+            kwargs (dict, optional): named arguments and their values use with the
```
Comment (daltonbohning) [inline src/tests/ftest/util/server_utils.py:327]: minor - 
```suggestion
            kwargs (dict, optional): named arguments and their values to use with the
```

File: src/tests/ftest/util/server_utils.py [hunk 8]
```diff
@@ -345,6 +348,10 @@ def scm_prepare(self, **kwargs):
     def scm_reset(self, **kwargs):
         """Run daos_server scm reset on the server hosts.
 
+        Args:
+            kwargs (dict, optional): named arguments and their values use with the
```
Comment (daltonbohning) [inline src/tests/ftest/util/server_utils.py:352]:
```suggestion
            kwargs (dict, optional): named arguments and their values to use with the
```

File: src/tests/ftest/util/server_utils.py [hunk 9]
```diff
@@ -361,6 +368,10 @@ def scm_reset(self, **kwargs):
     def nvme_prepare(self, **kwargs):
         """Run daos_server nvme prepare on the server hosts.
 
+        Args:
+            kwargs (dict, optional): named arguments and their values use with the
```
Comment (daltonbohning) [inline src/tests/ftest/util/server_utils.py:372]:
```suggestion
            kwargs (dict, optional): named arguments and their values to use with the
```

**Review by daltonbohning — APPROVED** [review]

File: src/tests/ftest/util/server_utils.py (hunk 8 above)
Comment (phender) [reply src/tests/ftest/util/server_utils.py]: Done

File: src/tests/ftest/util/server_utils.py (hunk 7 above)
Comment (phender) [reply src/tests/ftest/util/server_utils.py]: Done

File: src/tests/ftest/util/server_utils.py (hunk 9 above)
Comment (phender) [reply src/tests/ftest/util/server_utils.py]: Done

**Review by daltonbohning — DISMISSED** [review]

**Comment by phender** [comment]
Sample execution of new code:
```
2023/03/02 09:03:39 DEBUG                 _cleanup_procs: --------------------------------------------------------------------------------
2023/03/02 09:03:39 DEBUG                 _cleanup_procs: Cleaning up running processes after running ./harness/advanced.py
2023/03/02 09:03:39 DEBUG                 _cleanup_procs: Looking for running processes: daos_server|daos_engine|daos_agent|cart_ctl|orterun|mpirun|dfuse
2023/03/02 09:03:39 DEBUG                 stop_processes: Searching for any processes on wolf-105vm[1-4] that match 'daos_server|daos_engine|daos_agent|cart_ctl|orterun|mpirun|dfuse'
2023/03/02 09:03:39 DEBUG                     run_remote: Running on wolf-105vm[1-4] with a 60 second timeout: /usr/bin/pgrep --list-full 'daos_server|daos_engine|daos_agent|cart_ctl|orterun|mpirun|dfuse'
2023/03/02 09:03:39 DEBUG                     log_output:   wolf-105vm4 (rc=0): 335521 /usr/bin/dfuse /tmp/daos_dfuse_test_cleanup_procs TestLabel_1 122E0C88-6A46-4372-BA58-260E1EF604AD
2023/03/02 09:03:39 DEBUG                     log_output:   wolf-105vm[1-3] (rc=1): <NONE>
2023/03/02 09:03:39 DEBUG                 stop_processes: Killing any processes on wolf-105vm4 that match 'daos_server|daos_engine|daos_agent|cart_ctl|orterun|mpirun|dfuse' and then waiting 5 seconds
2023/03/02 09:03:39 DEBUG                     run_remote: Running on wolf-105vm4 with a 60 second timeout: sudo /usr/bin/pkill 'daos_server|daos_engine|daos_agent|cart_ctl|orterun|mpirun|dfuse'
2023/03/02 09:03:39 DEBUG                     log_output:   wolf-105vm4 (rc=0): <NONE>
2023/03/02 09:03:44 DEBUG                     run_remote: Running on wolf-105vm4 with a 60 second timeout: /usr/bin/pgrep --list-full 'daos_server|daos_engine|daos_agent|cart_ctl|orterun|mpirun|dfuse'
2023/03/02 09:03:44 DEBUG                     log_output:   wolf-105vm4 (rc=1): <NONE>
2023/03/02 09:03:44 DEBUG                 stop_processes: All processes running on wolf-105vm4 that match 'daos_server|daos_engine|daos_agent|cart_ctl|orterun|mpirun|dfuse' have been stopped.
2023/03/02 09:03:44 WARNING                   _warn_test: Running processes found on wolf-105vm4
```

File: src/tests/ftest/util/general_utils.py (hunk 4 above)
Comment (phender) [reply src/tests/ftest/util/general_utils.py]: Done and code has been moved to src/tests/ftest/util/run_utils.py

File: src/tests/ftest/cart/util/cart_utils.py [hunk 10]
```diff
@@ -158,12 +158,10 @@ def cleanup_processes(self):
...
-                retry_count += 1
-            else:
-                self.log.info("All '%s' processes have been stopped", processes)
-                retry_count = 99
+            _, running = stop_processes(self.log, localhost, processes, exclude=negative_filter)
+            if not running:
+                retry_count = 2
+            retry_count += 1
         if retry_count == 2:
             error_list.append("Unable to stop cart processes!")
```
Comment (daltonbohning) [inline src/tests/ftest/cart/util/cart_utils.py:166]: Pre-existing, but I think the usage of `retry_count` is confusing here. Personally, I would do something like:
```
        running = True
        for _ in range(2):
            _, running = stop_processes(self.log, localhost, processes, exclude=negative_filter)
            if not running:
                break
        if running
            error_list.append("Unable to stop cart processes!")
```

File: src/tests/ftest/launch.py [hunk 11]
```diff
@@ -2411,19 +2412,13 @@ def _cleanup_procs(self, test):
...
-            logger.debug("Killing running processes: %s", proc_pattern)
-            pkill_cmd = f"sudo -n pkill -e --signal KILL '{proc_pattern}'"
-            pkill_result = run_remote(logger, pgrep_result.passed_hosts, pkill_cmd)
-            if pkill_result.failed_hosts:
-                message = f"Failed to kill processes on {pkill_result.failed_hosts}"
-                self._fail_test(self.result.tests[-1], "Process", message)
-            else:
-                message = f"Running processes found on {pgrep_result.passed_hosts}"
-                self._warn_test(self.result.tests[-1], "Process", message)
+        detected, running = stop_processes(logger, hosts, f"'{proc_pattern}'")
```
Comment (daltonbohning) [inline src/tests/ftest/launch.py:2415]: As you asked, it wouldn't hurt to use `force=True` here, but this is already more robust than previous behavior

File: src/tests/ftest/util/run_utils.py [hunk 12]
```diff
@@ -368,3 +369,70 @@ def find_command(source, pattern, depth, other=None):
...
+            hosts the processes matching the pattern are still running (will be empty if every
+            process was killed or no process matching the pattern were found).
+
+    """
+    processes_detected = NodeSet()
+    processes_running = NodeSet()
+    command = f"/usr/bin/pgrep --list-full {pattern}"
+    pattern_match = str(pattern)
+    if exclude:
+        command = f"/usr/bin/ps xa | grep -E {pattern} | grep -vE {exclude}"
```
Comment (daltonbohning) [inline src/tests/ftest/util/run_utils.py:400]: We should probably quote the args to be safe
```suggestion
        command = f"/usr/bin/ps xa | grep -E '{pattern}' | grep -vE '{exclude}'"
```

File: src/tests/ftest/util/run_utils.py [hunk 13]
```diff
@@ -368,3 +369,70 @@ def find_command(source, pattern, depth, other=None):
...
+    # Attempt to kill any processes found on any of the hosts with increasing force
+    steps = [("", 5), (" --signal ABRT", 1), (" --signal KILL", 0)]
+    if force:
+        steps = [(" --signal KILL", 5)]
+    while steps and result.passed_hosts:
+        step = steps.pop(0)
+        log.debug(
+            "Killing%s any processes on %s that match %s and then waiting %s seconds",
+            step[0], result.passed_hosts, pattern_match, step[1])
+        kill_command = f"sudo /usr/bin/pkill{step[0]} {pattern}"
```
Comment (daltonbohning) [inline src/tests/ftest/util/run_utils.py:425]: Suggesting quotes
```suggestion
        kill_command = f"sudo /usr/bin/pkill{step[0]} '{pattern}'"
```

**Review by daltonbohning — DISMISSED** [review]

**Review by daltonbohning — APPROVED** [review]

**Comment by phender** [comment]
Failures in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11550/15/

- Known issue https://daosio.atlassian.net/browse/DAOS-12829 
  - Test / Functional on EL 8 / FTEST_pool.DynamicServerPool.1-./pool/dynamic_server_pool.py:DynamicServerPool.test_dynamic_server_pool;run-extra_servers-hosts-pool-server_config-engines-0-storage-0-c233

**Comment by phender** [comment]
All tests passed in https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11550/16/

**Review by dinghwah — APPROVED** [review]

**Review by sylviachanoiyee — APPROVED** [review]

**Pull request merged**